doc/man/gnunet-vpn.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd February 25, 2012
.Dt GNUNET-VPN 1
.Os
.Sh NAME
.Nm gnunet-vpn
.Nd manually setup a GNUnet VPN tunnel
.Sh SYNOPSIS
.Nm
.Op Fl 4 | -ipv4
.Op Fl 6 | -ipv6
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl d Ar TIME | Fl -duration Ar TIME
.Op Fl h | -help
.Op Fl i Ar IP | Fl -ip Ar IP
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl p Ar PEERID | Fl -peer= Ns Ar PEERID
.Op Fl s Ar NAME | Fl -service= Ns Ar NAME
.Op Fl t | -tcp
.Op Fl u | -udp
.Op Fl V | -verbose
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
can be used to manually setup a VPN tunnel via the GNUnet network.
There are two main types of tunnels.
Tunnels to an exit node which routes the traffic to the global Internet, and tunnels to a node that runs a service only within GNUnet.
Depending on the type of tunnel, gnunet-vpn takes different options.
The
.Fl i
option is required for tunnels to an exit node, whereas the
.Fl p
and
.Fl s
options in conjunction with either
.Fl u
or
.Fl t
are required for tunnels to services.
For exit tunnels, both UDP and TCP traffic will be redirected.
For service tunnels, either UDP
.Pq Fl u
or TCP
.Pq Fl t
traffic will be redirected.
The tool will display the IP address for this end of the tunnel.
The address can be displayed as soon as it has been allocated, or only after
.Pq Fl a
the tunnel has been created.
.Bl -tag -width indent
.It Fl 4 | -ipv4
Desired IP address on this end of the tunnel should be an IPv4 address.
.It Fl 6 | -ipv6
Desired IP address on this end of the tunnel should be an IPv6 address.
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl d Ar TIME | Fl -duration Ar TIME
The mapping should be established for TIME.
The value given must be a number followed by a space and a time unit, for example "500 ms".
Note that the quotes are required on the shell.
The default vaue is 5 minutes.
.It Fl h | -help
Print short help on options.
.It Fl i Ar IP | Fl -ip Ar IP
Tunnel should be to an exit node and connect to the given IPv4 or IPv6
IP address.
Note that you can specify an IPv6 address as the target here, even in
combination with "-4" (4to6) and similarly you can specify an IPv4
address in combination with "-6" (6to4).
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl p Ar PEERID | Fl -peer= Ns Ar PEERID
Name of the peer offering the service to connect to.
Cannot be used in conjunction with
.Fl i ,
requires
.Fl s .
.It Fl s Ar NAME | Fl -service= Ns Ar NAME
Name of the service running on the target peer.
Cannot be used in conjunction with
.Fl i ,
requires
.Fl p .
.It Fl t | -tcp
Service runs TCP.
Either
.Fl t
or
.Fl u
must be specified when using
.Fl s .
.It Fl u | -udp
Service runs UDP.
Either
.Fl t
or
.Fl u
must be specified when using
.Fl s .
.It Fl V | -verbose
Be verbose.
.It Fl v | -version
Print GNUnet version number.
.El
.Sh SEE ALSO
.Xr gnunet-setup 1
.sp
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Thu Oct 17 00:00:00 UTC 2019
  Added support for doas, use it in some places conditionally
  if sudo is not present. Fixed sudo detection. -ng0

Mon Oct 14 00:00:00 UTC 2019
  Added 'pretty' make rule to run uncrustify over the source tree.
  -ng0

Mon Sep 16 00:00:00 UTC 2019
  Remove checks for CXX and CPP. Scans found no
  (remaining) cxx code. ccache with clang on
  NetBSD 9 hit no rebuilds. -ng0

Mon Sep 10 00:00:00 UTC 2019
  Drop win32 and mingw support after discussion with
  Christian. It has been unmaintained in our code base
  for years and there are no known users.
  Future implementors for Windows 10 support can
  refer to this commit as reference. -ng0

Mon Sep 9 00:00:00 UTC 2019
  Remove plibc, closes #5877 as discussed in
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00002.html
  -ng0

Sun Sep 1 13:25:07 2019 +0200
  Improved support for external applications/services wrt
  localization/gettext.
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00000.html

Tue 20 Aug 2019 04:32:09 PM CEST
  Fixed fatal bug in gnunet-bcd that prevented the main page
  from showing up. -CG

Mon Aug 19 17:36:48 2019 +0200
  REST: Added API plugin to read/write GNUnet configuration options -alexia

Sun Jul 28 00:00:00 2019 +0000
  Adding --enable-texinfo4 configure switch, defaulting to off.
  This allows distributors to override the texinfo 5.x and later
  commandline options we use, and allows us on the serverside to
  build html with newer output options. -ng0

Wed 24 Jul 2019 05:43:43 PM CEST
  Releasing GNUnet 0.11.6. -CG

Mon Jul 15 09:57:20 2019 +0200
  The REST server now has a configurable CORS Origin setting -schanzen

Fri Jul 5 22:11:04 2019 +0200
  Added detection of SUID binaries
  through GNUNET_OS_get_suid_binary_path -lurchi

Sat 29 Jun 2019 10:18:55 AM CEST
  Renamed "heap" namestore backend to "flat" as it does
  persist data in a file. Made it work with databases larger
  than 16 MB. -CG

Fri Jun 28 14:10:53 2019 +0200
  RECLAIM: Reissue of tickets (same attributes, same identity)
  will no longer create a new ticket, but reuse the old. -alexia

Thu Jun 27 10:34:46 2019 +0200
  The REST server now returns the Origin header value in the
  CORS header reponse. -schanzen

Thu Jun 27 09:08:06 2019 +0200
  GNS NSS plugin will not timeout and not hang forever and block
  gethostbyname. Final resolution pending #5782 -schanzen

Sat Jun 22 10:38:43 2019 +0200
  Add gnutls support for GNS CA certificate generation -schanzen

Sat 15 Jun 2019 11:45:35 PM CEST
  Improved gnunet-gns performance for many ego scenario. -CG

Fri 14 Jun 2019 07:17:40 PM CEST
  Add option to gnunet-identity to display private keys. -CG

Wed Jun 5 21:19:46 2019 +0200
 Releasing GNUnet 0.11.5. -schanzen

Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression
  in 0.11.0 where documentation was an on/off switch.
  New introduced: --enable-include-manpages.
  Fixed: texi2mdoc integration (--enable-texi2mdoc-generation).
  buildsystem: decrease python dependency to 3.4 from
  previously 3.7.
  -ng0

Tue Mar 12 10:06:17 CET 2019
  REST: Move back from rest-plugin into respective subsystem directories.

Mon Mar 11 02:47:55 CET 2019
  Make configure complain if no working libidn is detected. -CG

Sat Mar 9 15:58:45 2019 +0100
  REST: Config option for address bind. Defaults to localhost.
  GNS: dns2gns/gns-proxy config option for address bind. Defaults to localhost.

Sat Mar 9 01:58:22 CET 2019
  gnunet-publish now by default does not expose the creation time,
  option -d no longer exists, to enable the old behavior use -E. -CG

Sat Mar 2 14:23:10 2019 +0100

src/vpn/gnunet-vpn.c
/*
     This file is part of GNUnet.
     Copyright (C) 2012 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file src/vpn/gnunet-vpn.c
 * @brief Tool to manually request VPN tunnels to be created
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_util_lib.h"
#include "gnunet_tun_lib.h"
#include "gnunet_vpn_service.h"


/**
 * Handle to vpn service.
 */
static struct GNUNET_VPN_Handle *handle;

/**
 * Opaque redirection request handle.
 */
static struct GNUNET_VPN_RedirectionRequest *request;

/**
 * Option -p: destination peer identity for service
 */
static char *peer_id;

/**
 * Option -s: service name (hash to get service descriptor)
 */
static char *service_name;

/**
 * Option -i: target IP
 */
static char *target_ip;

/**
 * Option -4: IPv4 requested.
 */
static int ipv4;

/**
 * Option -6: IPv6 requested.
 */
static int ipv6;

/**
 * Option -t: TCP requested.
 */
static int tcp;

/**
 * Option -u: UDP requested.
 */
static int udp;

/**
 * Selected level of verbosity.
 */
static unsigned int verbosity;

/**
 * Global return value.
 */
static int ret;

/**
 * Option '-d': duration of the mapping
 */
static struct GNUNET_TIME_Relative duration = { 5 * 60 * 1000 };


/**
 * Shutdown.
 */
static void
do_disconnect (void *cls)
{
  if (NULL != request)
  {
    GNUNET_VPN_cancel_request (request);
    request = NULL;
  }
  if (NULL != handle)
  {
    GNUNET_VPN_disconnect (handle);
    handle = NULL;
  }
  GNUNET_free_non_null (peer_id);
  GNUNET_free_non_null (service_name);
  GNUNET_free_non_null (target_ip);
}


/**
 * Callback invoked from the VPN service once a redirection is
 * available.  Provides the IP address that can now be used to
 * reach the requested destination.
 *
 * @param cls closure
 * @param af address family, AF_INET or AF_INET6; AF_UNSPEC on error;
 *                will match 'result_af' from the request
 * @param address IP address (struct in_addr or struct in_addr6, depending on 'af')
 *                that the VPN allocated for the redirection;
 *                traffic to this IP will now be redirected to the
 *                specified target peer; NULL on error
 */
static void
allocation_cb (void *cls, int af, const void *address)
{
  char buf[INET6_ADDRSTRLEN];

  request = NULL;
  switch (af)
  {
  case AF_INET6:
  case AF_INET:
    fprintf (stdout, "%s\n", inet_ntop (af, address, buf, sizeof(buf)));
    break;

  case AF_UNSPEC:
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR, _ ("Error creating tunnel\n"));
    ret = 1;
    break;

  default:
    break;
  }
  GNUNET_SCHEDULER_shutdown ();
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  int dst_af;
  int req_af;
  struct GNUNET_PeerIdentity peer;
  struct GNUNET_HashCode sd;
  const void *addr;
  struct in_addr v4;
  struct in6_addr v6;
  uint8_t protocol;
  struct GNUNET_TIME_Absolute etime;

  etime = GNUNET_TIME_relative_to_absolute (duration);
  GNUNET_SCHEDULER_add_shutdown (&do_disconnect, NULL);
  handle = GNUNET_VPN_connect (cfg);
  if (NULL == handle)
    goto error;
  req_af = AF_UNSPEC;
  if (ipv4)
  {
    if (ipv6)
    {
      fprintf (stderr,
               _ ("Option `%s' makes no sense with option `%s'.\n"),
               "-4",
               "-6");
      goto error;
    }
    req_af = AF_INET;
  }
  if (ipv6)
    req_af = AF_INET6;

  if (NULL == target_ip)
  {
    if (NULL == service_name)
    {
      fprintf (stderr, _ ("Option `%s' or `%s' is required.\n"), "-i", "-s");
      goto error;
    }
    if (NULL == peer_id)
    {
      fprintf (stderr,
               _ ("Option `%s' is required when using option `%s'.\n"),
               "-p",
               "-s");
      goto error;
    }
    if (! (tcp | udp))
    {
      fprintf (stderr,
               _ ("Option `%s' or `%s' is required when using option `%s'.\n"),
               "-t",
               "-u",
               "-s");
      goto error;
    }
    if (tcp & udp)
    {
      fprintf (stderr,
               _ ("Option `%s' makes no sense with option `%s'.\n"),
               "-t",
               "-u");
      goto error;
    }
    if (tcp)
      protocol = IPPROTO_TCP;
    if (udp)
      protocol = IPPROTO_UDP;
    if (GNUNET_OK !=
        GNUNET_CRYPTO_eddsa_public_key_from_string (peer_id,
                                                    strlen (peer_id),
                                                    &peer.public_key))
    {
      fprintf (stderr, _ ("`%s' is not a valid peer identifier.\n"), peer_id);
      goto error;
    }
    GNUNET_TUN_service_name_to_hash (service_name, &sd);
    request = GNUNET_VPN_redirect_to_peer (handle,
                                           req_af,
                                           protocol,
                                           &peer,
                                           &sd,
                                           etime,
                                           &allocation_cb,
                                           NULL);
  }
  else
  {
    if (1 != inet_pton (AF_INET6, target_ip, &v6))
    {
      if (1 != inet_pton (AF_INET, target_ip, &v4))
      {
        fprintf (stderr, _ ("`%s' is not a valid IP address.\n"), target_ip);
        goto error;
      }
      else
      {
        dst_af = AF_INET;
        addr = &v4;
      }
    }
    else
    {
      dst_af = AF_INET6;
      addr = &v6;
    }
    request = GNUNET_VPN_redirect_to_ip (handle,
                                         req_af,
                                         dst_af,
                                         addr,
                                         etime,
                                         &allocation_cb,
                                         NULL);
  }
  return;

error:
  GNUNET_SCHEDULER_shutdown ();
  ret = 1;
}


int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] =
  { GNUNET_GETOPT_option_flag ('4',
                               "ipv4",
                               gettext_noop (
                                 "request that result should be an IPv4 address"),
                               &ipv4),

    GNUNET_GETOPT_option_flag ('6',
                               "ipv6",
                               gettext_noop (
                                 "request that result should be an IPv6 address"),
                               &ipv6),

    GNUNET_GETOPT_option_relative_time (
      'd',
      "duration",
      "TIME",
      gettext_noop ("how long should the mapping be valid for new tunnels?"),
      &duration),

    GNUNET_GETOPT_option_string ('i',
                                 "ip",
                                 "IP",
                                 gettext_noop (
                                   "destination IP for the tunnel"),
                                 &target_ip),

    GNUNET_GETOPT_option_string (
      'p',
      "peer",
      "PEERID",
      gettext_noop ("peer offering the service we would like to access"),
      &peer_id),

    GNUNET_GETOPT_option_string ('s',
                                 "service",
                                 "NAME",
                                 gettext_noop (
                                   "name of the service we would like to access"),
                                 &service_name),

    GNUNET_GETOPT_option_flag ('t',
                               "tcp",
                               gettext_noop ("service is offered via TCP"),
                               &tcp),

    GNUNET_GETOPT_option_flag ('u',
                               "udp",
                               gettext_noop ("service is offered via UDP"),
                               &udp),

    GNUNET_GETOPT_option_verbose (&verbosity),

    GNUNET_GETOPT_OPTION_END };

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;

  ret =
    (GNUNET_OK == GNUNET_PROGRAM_run (argc,
                                      argv,
                                      "gnunet-vpn",
                                      gettext_noop ("Setup tunnels via VPN."),
                                      options,
                                      &run,
                                      NULL))
    ? ret
    : 1;
  GNUNET_free ((void *) argv);
  return ret;
}


/* end of gnunet-vpn.c */
